pep-0484.txt

```diff
@@ -276,6 +276,34 @@ implemented by deferring to ``isinstance(x, collections.abc.Callable)``.
 However, ``isinstance(x, typing.Callable[...])`` is not supported.
 
 
+Decorators
+----------
+
+Decorators can modify the types of the functions or classes they
+decorate. Use the ``decorated_type`` decorator to declare the type of
+the resulting item after all other decorators have been applied::
+
+  from typing import ContextManager, Iterator, decorated_type
+  from contextlib import contextmanager
+
+  class DatabaseSession: ...
+
+  @decorated_type(Callable[[str], ContextManager[DatabaseSession]])
+  @contextmanager
+  def session(url: str) -> Iterator[DatabaseSession]:
+      s = DatabaseSession(url)
+      try:
+          yield s
+      finally:
+          s.close()
+
+The argument of ``decorated_type`` is a type annotation on the name
+being declared (``session``, in the example above). If you have
+multiple decorators, ``decorated_type`` must be topmost.  The
+``decorated_type`` decorator is invalid on a function declaration that
+is also decorated with ``overload``, but you can annotate the
+implementation of the overload series with ``decorated_type``.
```

Comment:
Should we clarify (via example and/or specification language) that the type in `@decorated_type` may differ from the naturally inferred type, and that the inferred type must be assignable to the declared type? (Like in assignments or passing an argument to a function.) E.g. either one may have an `Any` where the other has something more specific.

I'm not sure if the semantics should exactly match those of assignment, though that would be my first approximation until we find a counterexample.